specs/CustomDownload.md

```diff
@@ -0,0 +1,775 @@
+    - Uri
+    - Mime type
+    - Content disposition
+    - Download size in bytes
+    - Result file path (get)
+    - Progress size in bytes
+    - Estimated end time
+    - State (in progress, completed, interrupted)
+    - Pause/resume/cancel
+    - Deferral
+    - Events:
+        - DownloadStateChanged
+            - InterruptReason
+        - DownloadProgressSizeInBytesChanged
+        - DownloadEstimatedEndTimeChanged
+3. WebView Setting: Default Download Dialog - used to disable the default download UI (bottom dock)
+
+We believe these 3 parts of a redesigned Download API, should enable your app have a customizable & complete download experience.
+
+# Examples
+<!-- TEMPLATE
+    Use this section to explain the features of the API, showing
+    example code with each description in both C# (for our WinRT API or .NET API) and
+    in C++ for our COM API. Use snippets of the sample code you wrote for the sample apps.
+
+    The general format is:
+
+    ## FirstFeatureName
+
+    Feature explanation text goes here, including why an app would use it, how it
+    replaces or supplements existing functionality.
+
+    ```c#
+    void SampleMethod()
+    {
+        var show = new AnExampleOf();
+        show.SomeMembers = AndWhyItMight(be, interesting)
+    }
+    ```
+
+    ```cpp
+    void SampleClass::SampleMethod()
+    {
+        winrt::com_ptr<ICoreWebView2> webview2 = ...
+    }
+    ```
+
+    ## SecondFeatureName
+
+    Feature explanation text goes here, including why an app would use it, how it
+    replaces or supplements existing functionality.
+
+    ```c#
+    void SampleMethod()
+    {
+        var show = new AnExampleOf();
+        show.SomeMembers = AndWhyItMight(be, interesting)
+    }
+    ```
+
+    ```cpp
+    void SampleClass::SampleMethod()
+    {
+        winrt::com_ptr<ICoreWebView2> webview2 = ...
+    }
+    ```
+
+    As an example of this section, see the Examples section for the PasswordBox
+    control (https://docs.microsoft.com/windows/uwp/design/controls-and-patterns/password-box#examples).
+-->
+## Win32 C++
+```cpp
+// Register a handler for the `DownloadStarting` event.
+// This example hides the default download dialog and shows a dialog box instead.
+// The dialog box displays the default save path and allows the user to specify a different path.
+// Selecting `OK` will save the download to the chosen path.
+// Selecting `CANCEL` will cancel the download.
+CHECK_FAILURE(m_webView->add_DownloadStarting(
+    Callback<ICoreWebView2DownloadStartingEventHandler>(
+        [this](
+            ICoreWebView2_3* sender,
+            ICoreWebView2DownloadStartingEventArgs* args) -> HRESULT {
+            auto showDialog = [this, args] {
+
+                // Hide the default download dialog.
+                wil::com_ptr<ICoreWebView2Settings> m_settings;
+                CHECK_FAILURE(m_webView->get_Settings(&m_settings));
+                CHECK_FAILURE(
```

Comment:
The ShouldDisplayDefaultDownloadDialog setting applies to the entire webview2. You probably want to set this once outside of your event handler correct not again and again inside your event handler? Also not sure if its too late to set this inside the event handler?